databricks/koalas/frame.py

```diff
@@ -7830,43 +7830,89 @@ def reindex(
 
     def _reindex_index(self, index, fill_value):
         # When axis is index, we can mimic pandas' by a right outer join.
-        assert (
-            len(self._internal.index_spark_column_names) <= 1
-        ), "Index should be single column or not set."
+        nlevels = len(self._internal.index_spark_column_names)
+        assert nlevels <= 1 or (
+            isinstance(index, ks.MultiIndex) and nlevels == index.nlevels
+        ), "MultiIndex DataFrame can only be reindexed with a similar koalas MultiIndex ."
 
-        index_column = self._internal.index_spark_column_names[0]
-
-        kser = ks.Series(list(index))
-        labels = kser._internal.spark_frame.select(kser.spark.column.alias(index_column))
+        index_columns = self._internal.index_spark_column_names
         frame = self._internal.resolved_copy.spark_frame.drop(NATURAL_ORDER_COLUMN_NAME)
 
+        if isinstance(index, ks.Index):
+            if nlevels != index.nlevels:
+                # Use temporary column so we can call reindex in order to fill columns on new index
+                return DataFrame(
+                    index._internal.with_new_columns(
+                        [F.lit(0)],
+                        column_labels=[(verify_temp_column_name(frame, "__temp_spark_column__"),)],
+                    )
+                ).reindex(columns=self._internal.data_spark_column_names, fill_value=fill_value)
+
+            obj = index
+            index_map = OrderedDict(
+                [(SPARK_INDEX_NAME_FORMAT(i), (name,)) for i, name in enumerate(index.names)]
+            )
+            labels = obj._internal.spark_frame.select(index_columns)
+        else:
+            obj = ks.Series(list(index))
+            index_map = self._internal._index_map
+            scols = obj._internal.data_spark_columns
+            labels = obj._internal.spark_frame.select(
+                [scols[i].alias(index_columns[i]) for i in range(len(scols))]
+            )
+
         if fill_value is not None:
-            frame_index_column = verify_temp_column_name(frame, "__frame_index_column__")
-            frame = frame.withColumnRenamed(index_column, frame_index_column)
+            frame_index_columns = [
+                verify_temp_column_name(frame, "__frame_index_column_{}__".format(i))
+                for i in range(nlevels)
+            ]
+            cols = [F.col(t[0]).alias(t[1]) for t in zip(index_columns, frame_index_columns)]
```

Comment:
`scol_for(frame, t[0])` instead of `F.col(t[0])`?

Reply:
Btw, we prefer to use a tuple, e.g., `[... for index_col, frame_index_col in zip(index_columns, frame_index_columns)]`.